665 consolidate emails

esoterik opened the pull request:
Resolves #665 and #938

orenyk commented on app/views/app_configs/_form.erb:
```diff
@@ -59,6 +59,7 @@
     <%= f.input :enable_renewals, label: 'Enable renewals', hint: 'Allow users to renew reservations.' %>
   </fieldset>
 
+<<<<<<< HEAD
```
this looks leftover from a merge conflict :-P

orenyk commented on app/views/user_mailer/reservation_status_update.html.erb:
```diff
@@ -0,0 +1,30 @@
+<p><%= link_to 'Reservation Information', reservation_url(@reservation, only_path: false) %></p>
+
+<p>Equipment: <%= @reservation.equipment_model.name%> 
+<% if @reservation.equipment_object then %> (<%= @reservation.equipment_object.name %>) <% end %></p>
```
I don't think you need the `then` here

orenyk commented on app/models/reservation.rb:
```diff
@@ -77,7 +80,10 @@ def status # rubocop:disable CyclomaticComplexity, PerceivedComplexity
       end
     elsif checked_in.nil?
-      due_date < Time.zone.today ? 'overdue' : 'checked out'
+      if due_date.midnight == Date.current.midnight then 'due today'
```
This doesn't need the `.midnight` method, you can just do `due_date == Time.zone.today`

orenyk commented on app/models/reservation.rb:
```diff
@@ -77,7 +80,10 @@ def status # rubocop:disable CyclomaticComplexity, PerceivedComplexity
     elsif checked_in.nil?
-      due_date < Time.zone.today ? 'overdue' : 'checked out'
+      if due_date.midnight == Date.current.midnight then 'due today'
+      elsif due_date < Date.current then 'overdue'
```
similarly, `due_date < Time.zone.today`

orenyk commented on app/models/reservation.rb:
```diff
@@ -201,7 +207,7 @@ def checkin(checkin_handler, procedures, new_notes)
     if checked_in.to_date > due_date
       # equipment was overdue, send an email confirmati
```
mind fixing this comment while you're in here?

orenyk commented on app/controllers/reservations_controller.rb:
```diff
@@ -382,7 +382,7 @@ def checkout_email
   end
 
   def checkin_email
```
did you forget to update the `checkout_email` method?

orenyk replied:
These methods nominally get called in the `current_reservations`, `receipt`, and reservation `show` views, but have been commented out for some time. I think it's time we re-enable them :smile:. Note that we'll have to update the routes as well.

esoterik replied:
this is done except it made a few tests fail and I'm not quite sure why, so I'm waiting to commit it

orenyk commented on config/schedule.rb:
```diff
@@ -38,6 +38,7 @@
   rake 'deny_missed_requests'
   rake 'delete_old_blackouts'
   rake 'email_missed_reservations'
+  rake 'email_checkout_reminder'
```
you should also update the `scheduler` task for Heroku, now that that's been merged into `master`.

orenyk commented on db/schema.rb:
```diff
@@ -52,6 +52,8 @@
     t.boolean  "enable_new_users",                                   default: true
     t.integer  "res_exp_time"
     t.boolean  "enable_guests",                                      default: true
+    t.boolean  "upcoming_checkout_email_active"
```
we should probably set this to `default: true` like the `upcoming_checkin_email_active` setting

orenyk replied:
This will update once you fix your migrations and run `rake db:migrate`

orenyk commented on lib/tasks/email_checkout_reminder.rake:
```diff
@@ -0,0 +1,19 @@
+desc 'Send email reminder about upcoming checkout'
```
remind me - did you put this in a separate file for testing purposes?

esoterik replied:
yup. it would be easy to refactor all the email tasks into the same file and testable though

orenyk replied:
Let's leave it for now, we can have a wish list issue open to write tests for all of the mailman rake tasks and consolidate them into a single file.

orenyk commented on lib/tasks/email_checkout_reminder.rake:
```diff
@@ -0,0 +1,19 @@
+  if AppConfig.first.upcoming_checkout_email_active?
+    # get all reservations that start today
+    upcoming_reservations =
+      Reservation.where('checked_out IS NULL and '\
```
let's use our scopes; this could be `Reservation.untouched.where('start_date = ?', Time.zone.today) instead. We could also technically use the `starts_on_days(Time.zone.today, Time.zone.today)` but I think the former is better.

orenyk commented on app/views/user_mailer/_requested.html.erb:
```diff
@@ -0,0 +1,3 @@
+<p>Your <%= link_to 'reservation request', reservation_url(@reservation, only_path: false) %> has been approved!<p>
+
+<p>You can pick up your <%= @reservation.equipment_model.name %> at <%= @reservation.start_date.to_date.to_s(:long) %>. Remember to bring it back on <%= @reservation.due_date.to_date.to_s(:long) %>!
```
you can take out all the `to_date`s, since the `start_date` and `due_date`s are both `Date`s.

orenyk commented on app/views/user_mailer/_reserved.html.erb:
```diff
@@ -0,0 +1,3 @@
+<p>Your <%= link_to 'reservation request', reservation_url(@reservation, only_path: false) %> has been approved!<p>
+
+<p>You can pick up your <%= @reservation.equipment_model.name %> at <%= @reservation.start_date.to_date.to_s(:long) %>. Remember to bring it back on <%= @reservation.due_date.to_date.to_s(:long) %>!
```
same here (no `to_date`)

orenyk commented on app/views/user_mailer/_returned_overdue.html.erb:
```diff
@@ -0,0 +1,4 @@
+<p>This is a notification for your late equipment checkin. </p>
+
+<p>Your late fee is <%= number_to_currency(@reservation.equipment_model.late_fee) %> per day, for a total of <%= number_to_currency(((@reservation.checked_in.to_date - @reservation.due_date.to_date )).to_i*(@reservation.equipment_model.late_fee))%>
```
take out the `to_date` on `due_date`, but not on `checked_in` since that's still a `DateTime`.

orenyk replied:
second note (but off-topic), we shouldn't be calculating the late fee in the mail body (or in the reservation view), I'll open up a new issue for that.

orenyk commented on app/views/user_mailer/reservation_status_update.html.erb:
```diff
@@ -0,0 +1,30 @@
+
+<p>Equipment: <%= @reservation.equipment_model.name%> 
+<% if @reservation.equipment_object %> (<%= @reservation.equipment_object.name %>) <% end %></p>
+<p>Start Date: <%= @reservation.start_date.strftime(" %B %e, %Y at %l:%M%P")%> </p>
```
there shouldn't be any time for the start and due dates, only for checkin and checkout

orenyk commented on app/views/user_mailer/reservation_status_update.html.erb:
```diff
@@ -0,0 +1,30 @@
+<p>Start Date: <%= @reservation.start_date.strftime(" %B %e, %Y at %l:%M%P")%> </p>
+<p>Due Date: <%= @reservation.due_date.strftime(" %B %e, %Y at %l:%M%P")%> </p>
+<% if @reservation.checked_out %>
+  <p>Checkout date: <%= @reservation.checked_out.strftime(" %B %e, %Y at %l:%M%P") %></p>
```
Let's rename this "Checked Out"

orenyk commented on app/views/user_mailer/reservation_status_update.html.erb:
```diff
@@ -0,0 +1,30 @@
+  <p>Checkout date: <%= @reservation.checked_out.strftime(" %B %e, %Y at %l:%M%P") %></p>
+<% end %>
+<% if @reservation.checked_in %>
+  <p>Return date: <%= @reservation.checked_in.strftime(" %B %e, %Y at %l:%M%P") %></p>
```
similarly, this could be called "Checked In"

orenyk commented on spec/lib/tasks/email_checkout_reminder_rake_spec.rb:
```diff
@@ -0,0 +1,31 @@
+
+  before(:each) do
+    @upcoming = FactoryGirl.create(:valid_reservation,
+                                   start_date: Date.current,
```
switch `Time.zone.today` for `Date.current` re: #939

orenyk commented on spec/mailers/user_mailer_spec.rb:
```diff
@@ -23,80 +23,99 @@
     ActionMailer::Base.deliveries = []
+    @res = FactoryGirl.create(:valid_reservation,
+                              reserver: reserver,
+                              start_date: Date.current + 1)
```
same here (`Time.zone.today`)

orenyk commented on spec/mailers/user_mailer_spec.rb:
```diff
@@ -23,80 +23,99 @@
+      @em = FactoryGirl.create(:equipment_model, late_fee: 0)
+      @res.update_attributes(FactoryGirl.attributes_for(:checked_in_reservation,
+                                                        :overdue))
+      @res.update_attributes(equipment_model_id: @em.id)
```
just FYI, you can also do `@res.update_attributes(equipment_model: @em)`

orenyk replied:
ignore my last comment :stuck_out_tongue:

orenyk commented on spec/models/reservation_spec.rb:
```diff
@@ -100,11 +100,23 @@
   end
 
   context 'when not checked out' do
```
change this to specify it's not starting to day as well

orenyk commented on db/seeds.rb:
```diff
@@ -44,6 +44,8 @@
 TOS_TEXT = File.read(File.join(DEFAULT_MSGS, 'tos_text'))
 UPCOMING_CHECKIN_RES_EMAIL = File.read(File.join(DEFAULT_MSGS,
                                                  'upcoming_checkin_email'))
+UPCOMING_CHECKOUT_RES_EMAIL = File.read(File.join(DEFAULT_MSGS,
```
You need to also assign this to the generated `AppConfig` object below (~L394), otherwise it isn't used.

orenyk commented on app/mailers/user_mailer.rb:
```diff
@@ -8,68 +8,21 @@ class UserMailer < ActionMailer::Base
+    mail(to: @reservation.reserver.email,
+         subject: '[Reservations] ' \
+                  "#{@reservation.equipment_model.name.capitalize} "\
+                  " Status Update: #{status_formatted}")
```
Hmmm, I'm not in love with this subject, it gets a little unwieldy. What about "[Reservations] `equipment model name` `status`". For example, "[Reservations] Nikon D90 Available Today", "[Reservations] iPad 2 Due Today", "[Reservations] Composite RCA Adapter Overdue". I changed "Starts Today" to "Available Today", but otherwise I think it works. What do you think?

orenyk commented on db/default_messages/upcoming_checkout_email:
```diff
@@ -0,0 +1,8 @@
+Dear @user@,\n\n
```
Note that these `\n`s don't actually get parsed correctly (not that it's critical because they're just seed data, but it's an issue nonetheless), both here and in the other default_messages. We should fix them while we're here. Also, note that there's a typo in the start of `upcoming_checkin_email` (`Return_date@` ==> `@return_date@`).

orenyk commented:
This looks really nice. I left a bunch of inline comments, the biggest of which is the refactoring of the two reservations controller methods (`checkin_email` and `checkout_email`) into one and the re-introduction of that functionality to our views. The only other comment I have is that I encountered an ERB warning in both `_due_today.html.erb:1` and `_starts_today.html.erb:1`:

```
_starts_today.html.erb:1: warning: encountered \r in the middle of line, treated as a mere space
```

Not sure what that's all about, see if it's something we need to deal with. Great work!

orenyk commented:
Also note that the query for the `send_upcoming_checkin_reminder` task in `mailmain.rake` has a typo (line 6). We need to add a space at the end of the line (after `and`) to make it work. Since we're fixing e-mails here I think it falls within scope.

esoterik commented:
I fixed most of the issues you pointed out in the comments. I've refactored the checkin/checkout email methods, but it's failing quite a few tests so I haven't committed it yet. The parsing issues in the email bodies also have yet to be corrected.

esoterik commented:
The tests no longer fail because of the refactored `send_receipt` method; tests still need to be written for the new method though.

esoterik commented:
also--it seems like the `\r` issue is from a windows line ending character in the default messages for those two emails

orenyk commented on db/default_messages/upcoming_checkin_email:
```diff
@@ -1,9 +1,5 @@
+  Hey there, you have equipment due! Please return the following items before 4pm on @return_date@.\n\n\
   @equipment_list@\n\n\
-  If you fail to return your equipment on time you will be subject to a late fee of @late_fee@ per day. If you have lost the item you may additioally have to pay a replacement fee of Replacement_fee@.\n\
+  If you fail to return your equipment on time you will be subject to a late fee of @late_fee@ per day. If you have lost the item you may additionally have to pay a replacement fee of @replacement_fee@.\n\
```
ha nice catch :-)

orenyk commented on db/default_messages/upcoming_checkout_email:
```diff
@@ -0,0 +1,4 @@
+  Hey there, you have equipment reserved today! Please checkout the following items today:\n\n\
```
Do we know where those `\r` characters or how to get rid of them? I thought that Git generally handled line endings, so I'm surprised it's there, particularly since these files were generated by @squidgetx (who runs Linux). Any ideas?

squidgetx replied:
very odd..

orenyk commented on db/migrate/20150206002111_add_upcoming_checkout_email_active_to_app_config.rb:
```diff
@@ -0,0 +1,5 @@
+class AddUpcomingCheckoutEmailActiveToAppConfig < ActiveRecord::Migration
+  def change
+    change_column_default :app_configs, :upcoming_checkout_email_active, true
```
Looks like you edited the wrong migration (see below).

orenyk commented on db/migrate/20150221192751_add_default_to_upcoming_checkout_email.rb:
```diff
@@ -0,0 +1,4 @@
+class AddDefaultToUpcomingCheckoutEmail < ActiveRecord::Migration
+  def change
```
Looks like you edited the wrong migration (see above).

orenyk commented:
Whoops, I found another issue. We broke the `ReservationsHelper#manage_reservations_btn` method since it relies on the `Reservation#status` method. We need to modify the helper conditionals so that the button shows up on reservations that start or are due today.

orenyk commented:
Also, line breaks in the app config settings are rendering improperly in the e-mail body:

![selection_051](https://cloud.githubusercontent.com/assets/901808/6317744/396e57cc-ba34-11e4-919d-20a4eb2e41fb.png)

orenyk commented:
Also, the checkout receipt fails when the reservation is due back today or is overdue but hasn't yet been returned (since the status is now "due today"). We might need to pass a `receipt` parameter into `UserMailer#reservation_status_update` to enforce the `checked out` status for those calls.

Note also that my [comment](https://github.com/YaleSTC/reservations/pull/1161#issuecomment-75422541) above also applies to the logic determining whether or not to display the "send receipt" button.

orenyk commented on app/controllers/reservations_controller.rb:
```diff
@@ -369,20 +369,8 @@ def current
-    if UserMailer.checkin_receipt(@reservation).deliver
+  def send_receipt
+    if UserMailer.reservation_status_update(@reservation).deliver
       redirect_to :back
```
This is no good since if we call this from the `receipt` method that follows the `checkout` or `checkin` methods, then it tries to run `checkout` again resulting in a `RecordNotFound` exception. Why don't we redirect to the reservation in both cases?

orenyk commented:
Nice job dealing with round 1, I left a few more inline comments and found a couple more issues while manually testing out this functionality (see above). Once those are addressed you should also play around with this branch and make sure everything works as expected (all buttons / links and all rake tasks). Hopefully we can get this done in tomorrow's meeting!

esoterik commented:
okay, everything is fixed and working locally, about to force push onto this branch to fix the git issues, let's hope it works

esoterik commented:
I think this is all set?

orenyk commented:
Thanks, I'll have to review this tomorrow but hopefully we'll be able to get it in as well. What was your ultimate solution to fixing all the Git weirdness?

esoterik commented:
I made a new branch locally, merged in 938, cherrypicked from 665, then force pushed to this branch haha

orenyk commented on app/mailers/user_mailer.rb:
```diff
@@ -8,68 +8,35 @@ class UserMailer < ActionMailer::Base
+      @status = 'due today'
+    end
+
+    if receipt && (@status == 'due today' || @status == 'overdue')
```
We should probably also check to make sure it's been checked out if the receipt parameter is true. An alternative way to structure this would be to write a second method `send_receipt` that runs that validation (e.g. `return unless reservation.checked_out`) and then calls `reservation_status_update(reservation, true)`.

On that note, what happens if it's `returned overdue`? As written, it's not called a "receipt", but we can fix that in the view below (I'll leave a comment there).

esoterik replied:
perhaps the name "receipt" for the variable is misleading--I only wanted to use it to force sending a checkout receipt if requested from the reservations controller; other cases should work without the additional parameter

orenyk replied:
No I understand that, what I'm saying is that someone could end up issuing a request to the `get_receipt` route, passing a reservation that isn't yet checked out and therefore it should fail. Unless I missed it, we're not currently checking for that anywhere (so if you end up calling `reservation_status_update(res, true)` and `res.checked_out == nil` the e-mail view will probably fail since it will try to call `strftime` on `nil`). Does that make more sense?

esoterik replied:
right yeah, that was in response to `returned overdue` not being called a receipt. I think we're on the same page

orenyk commented on app/views/user_mailer/_returned_overdue.html.erb:
```diff
@@ -0,0 +1,4 @@
+<p>This is a notification for your late equipment checkin. </p>
```
We should rewrite this to say "This is a receipt for your late equipment return." and go from there for cases when a receipt e-mail is requested. We should, however, only show the late fee statement if there is a late fee, so we'll need to add another conditional here (`<% if @reservation.equipment_model.late_fee > 0 %>` etc)

orenyk commented on app/views/user_mailer/reservation_status_update.html.erb:
```diff
@@ -0,0 +1,30 @@
+
+<br />
+
+<p><%= link_to 'Reservation Information', reservation_url(@reservation, only_path: false) %></p>
```
Let's actually make the link text "Reservation #ID" so it's clear that it goes the reservation page.

orenyk commented on lib/tasks/mailman.rake:
```diff
@@ -3,13 +3,13 @@ task send_upcoming_checkin_reminder: :environment do
+      Reservation.where('checked_out IS NOT NULL and checked_in IS NULL and '\
                           'due_date >= ? and due_date < ?',
                         Time.zone.today, Time.zone.today + 1.day)
     puts "Found #{upcoming_reservations.size} reservations due for checkin."\
```
can you update these to `Rails.logger.info` calls instead?

orenyk commented on spec/controllers/reservations_controller_spec.rb:
```diff
@@ -1407,12 +1407,26 @@
-    pending 'E-mails get sent'
+    context 'fails to send email' do
+      before do
+        allow(UserMailer).to receive_message_chain(
```
nice!

orenyk commented on spec/models/reservation_spec.rb:
```diff
@@ -99,7 +99,19 @@
+    it { expect(reservation).to_not be_eligible_for_renew }
+  end
+
+  context 'when upcoming' do
```
how are these two contexts different? More specifically, why doesn't the second assertion (L115) below fail?

esoterik replied:
this is probably leftover from when I was editing the status method to get 'due today' and 'starts today' emails; it shouldn't be necessary now, I'll just checkout the reservation spec from master

orenyk replied:
Yea, that's what I thought.

orenyk commented:
Once we've dealt with validating the receipt requests, I think this should be good to go (whew).

esoterik commented:
once travis finishes this should be good!

orenyk commented on app/mailers/user_mailer.rb:
```diff
@@ -8,68 +8,37 @@ class UserMailer < ActionMailer::Base
+    end
+
+    if checkout_receipt && (@status == 'due today' || @status == 'overdue') &&
+       @reservation.checked_out
```
Hmm, I almost feel like we should `return` here if it's not checked out. As I see it:

1. If `checkout_receipt` (or `receipt`, I actually prefer the latter since we made one controller method for both cases)
  1. If checked out and status weirdness then fix status
  2. If checked out and status is checked out then do nothing
  3. If not checked out, don't send the e-mail
  4. If checked in, do nothing

Does that make sense?

esoterik replied:
I think so, but just to double check before I push, here's what I have now:
```
    if @reservation.start_date == Time.zone.today &&
       @status == 'reserved'
      @status = 'starts today'
    elsif @reservation.due_date == Time.zone.today &&
          @status == 'checked out'
      @status = 'due today'
    end

    if receipt 
      # force checkout receipt to be sent
      if @status == 'due today' || @status == 'overdue' 
        @status = 'checked out'
      end
      if @status == 'checked out' && !@reservation.checked_out 
        return
      end
    end

```

orenyk commented on app/views/reservations/current_reservations.html.erb:
```diff
@@ -33,8 +33,10 @@
-<%# end %>
+<% if @reservation %>
+  <% if (@reservation.status == "checked out") or (@reservation.status == "overdue") %>
+    <%= link_to "Email checkout receipt", {:controller => :reservations, :action => "send_receipt", :id => @reservation}, :method => "get", :class => 'btn btn-default' %>
```
Can we update this hash syntax? Honestly, this almost feels like we just want a helper method for the e-mail receipt button, but that's beyond scope so don't worry about it!

orenyk commented:
Oh no! What's the merge conflict? In either case, you can just squash rebase, then rebase onto master to resolve it. Good work!
